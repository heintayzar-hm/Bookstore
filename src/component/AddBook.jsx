import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line react/prefer-stateless-function
class AddBook extends React.Component {
  constructor(props) {
    super();
    this.props = props;
    this.state = {
      author: '',
      title: '',
    };
  }

  changeHandler = (e) => {
    e.preventDefault();
    this.setState((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value,
    }));
  }

  submitHandler = (e) => {
    e.preventDefault();
    const { addBook } = this.props;
    const { title, author } = this.state;
    if (title.trim() && author.trim()) {
      addBook(title, author);
      this.setState({
        author: '',
        title: '',
      });
    }
  }

  render() {
    // eslint-disable-next-line no-unused-vars
    const { author, title } = this.state;
    return (
      <>
        <article data-testid="form-container">
          <h2>Add New Book</h2>
          <form onSubmit={this.submitHandler}>
            <input name="title" onChange={this.changeHandler} data-testid="book-title-input" type="text" placeholder="Book Title" value={title} required />
            <input name="author" onChange={this.changeHandler} data-testid="author-title-input" type="text" placeholder="Author" value={author} required />
            <button type="submit">Add Book</button>
          </form>
        </article>
      </>
    );
  }
}

AddBook.propTypes = {
  addBook: PropTypes.func.isRequired,
};
export default AddBook;
