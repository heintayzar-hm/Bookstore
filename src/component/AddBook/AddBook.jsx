import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import DropDown from './dropdown';

// eslint-disable-next-line react/prefer-stateless-function
class AddBook extends React.Component {
  constructor(props) {
    super();
    this.props = props;
    this.state = {
      author: '',
      title: '',
      category: { id: 1, name: 'Action' },
    };
  }

  changeHandler = (e) => {
    e.preventDefault();
    this.setState((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value,
    }));
  }

  dropdownChangeHandler = (value) => {
    this.setState((prevState) => ({
      ...prevState,
      category: value,
    }));
  }

  submitHandler = (e) => {
    e.preventDefault();
    const { addBook } = this.props;
    const { title, author, category } = this.state;
    if (title.trim() && author.trim()) {
      addBook(title, author, category.name);
      this.setState({
        author: '',
        title: '',
        category: '',
      });
    }
  }

  render() {
    // eslint-disable-next-line no-unused-vars
    const { author, title } = this.state;
    const { categories } = this.props;
    if (categories.length === 0) {
      return <></>;
    }
    return (
      <>
        <article data-testid="form-container">
          <h2 className="font-important tracking-wider text-[1.25rem] font-bold text-[#888] uppercase pb-3 ">Add New Book</h2>
          <form onSubmit={this.submitHandler} className=" gap-4 grid grid-cols-1 sm:grid-cols-12">
            <input className="dark:bg-black dark:text-white sm:col-span-3 p-2 px-3 border focus:outline-none rounded focus:border-blue border-solid tracking-widest text-brown" name="title" onChange={this.changeHandler} data-testid="book-title-input" type="text" placeholder="Book Title" value={title} required />
            <input className=" dark:bg-black dark:text-white sm:col-span-3 p-2 px-3 border focus:outline-none rounded focus:border-blue border-solid tracking-widest text-brown" name="author" onChange={this.changeHandler} data-testid="author-title-input" type="text" placeholder="Author" value={author} required />
            <div className="sm:col-span-4"><DropDown changeHandler={this.dropdownChangeHandler} currentSelectName={categories[0]} categories={categories} /></div>
            <button className="sm:col-span-2 p-3 bg-blue text-white rounded hover:bg-white hover:text-blue" type="submit">Add Book</button>
          </form>
        </article>
      </>
    );
  }
}

const mapStateToProps = (state) => ({
  // specify which state data to provide to the component
  categories: state.categories,
});

AddBook.propTypes = {
  addBook: PropTypes.func.isRequired,
  categories: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.any),
    PropTypes.string,
  ]).isRequired,
};
export default connect(mapStateToProps)(AddBook);
