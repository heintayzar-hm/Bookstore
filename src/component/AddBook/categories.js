const categories = [
  {
    id: 1,
    name: 'Action',
  },
  {
    id: 2,
    name: 'Fiction',
  },
  {
    id: 3,
    name: 'Adventure',
  },
  {
    id: 4,
    name: 'Romance',
  },
  {
    id: 5,
    name: 'Mystery',
  },
  {
    id: 6,
    name: 'Science fiction',
  },
  {
    id: 7,
    name: 'Thriller',
  },
  {
    id: 8,
    name: 'Fantasy',
  },
  {
    id: 9,
    name: 'Horror',
  },
  {
    id: 10,
    name: 'Self-help',
  },
  {
    id: 11,
    name: 'Business',
  },
  {
    id: 12,
    name: 'Cooking',
  },
  {
    id: 13,
    name: 'Travel',
  },
  {
    id: 14,
    name: 'Health',
  },
  {
    id: 15,
    name: 'Biography',
  },
  {
    id: 16,
    name: 'Memoir',
  },
  {
    id: 17,
    name: 'History',
  },
  {
    id: 18,
    name: 'Reference',
  },
  {
    id: 19,
    name: 'Textbook',
  },
  {
    id: 20,
    name: 'Poetry',
  },
  {
    id: 21,
    name: 'Religion',
  },
  {
    id: 22,
    name: 'Philosophy',
  },
  {
    id: 23,
    name: 'Parenting',
  },
  {
    id: 24,
    name: 'Crafts',
  },
  {
    id: 25,
    name: 'Art',
  },
];

export default categories;
