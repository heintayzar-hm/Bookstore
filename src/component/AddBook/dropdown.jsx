import { Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import { Listbox, Transition } from '@headlessui/react';
import { CheckIcon, ChevronUpDownIcon } from '@heroicons/react/20/solid';
import categories from './categories';

function classNames(...classes) {
  return classes.filter(Boolean).join(' ');
}

const DropDown = ({ changeHandler, currentSelectName = 'Adventure' }) => {
  const data = categories.filter((category) => category.name === currentSelectName);
  const [selected, setSelected] = useState(categories[(data[0].id) - 1]);

  const selectHandler = (currentSelected) => {
    setSelected(currentSelected);
    changeHandler(currentSelected);
  };

  return (
    <Listbox value={selected} onChange={selectHandler}>
      {({ open }) => (
        <>
          <div className="relative h-full">
            <Listbox.Button className="relative w-full h-full cursor-default rounded-md border border-gray-300 bg-white dark:bg-black py-2 pl-3 pr-10 text-left shadow-sm focus:border-blue focus:outline-none focus:ring-1 focus:ring-blue sm:text-sm">
              <span className="flex items-center">
                <span className="ml-3 block truncate">{selected.name}</span>
              </span>
              <span className="pointer-events-none absolute inset-y-0 right-0 ml-3 flex items-center pr-2">
                <ChevronUpDownIcon className="h-5 w-5 text-gray-400" aria-hidden="true" />
              </span>
            </Listbox.Button>

            <Transition
              show={open}
              as={Fragment}
              leave="transition ease-in duration-100"
              leaveFrom="opacity-100"
              leaveTo="opacity-0"
            >
              <Listbox.Options className="absolute top-auto bottom-0 z-10 mt-1 max-h-56 w-full overflow-auto rounded-md bg-white py-1 text-base shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none sm:text-sm">
                {categories.map((person) => (
                  <Listbox.Option
                    key={person.id}
                    className={({ active }) => classNames(
                      active ? 'text-white bg-blue' : 'text-gray-900 dark:text-white dark:bg-black',
                      'relative cursor-default select-none py-2 pl-3 pr-9',
                    )}
                    value={person}
                  >
                    {({ selected, active }) => (
                      <>
                        <div className="flex items-center">
                          <span
                            className={classNames(selected ? 'font-semibold' : 'font-normal', 'ml-3 block truncate')}
                          >
                            {person.name}
                          </span>
                        </div>

                        {selected ? (
                          <span
                            className={classNames(
                              active ? 'text-white' : 'text-blue',
                              'absolute inset-y-0 right-0 flex items-center pr-4',
                            )}
                          >
                            <CheckIcon className="h-5 w-5" aria-hidden="true" />
                          </span>
                        ) : null}
                      </>
                    )}
                  </Listbox.Option>
                ))}
              </Listbox.Options>
            </Transition>
          </div>
        </>
      )}
    </Listbox>
  );
};

DropDown.propTypes = {
  changeHandler: PropTypes.func.isRequired,
  currentSelectName: PropTypes.string.isRequired,
};
export default DropDown;
