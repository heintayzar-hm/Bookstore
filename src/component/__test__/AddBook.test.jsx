import { describe, test } from 'vitest';
import { render } from '@testing-library/react';
import AddBook from '../AddBook';

describe('AddBook', () => {
  test('renders the initial state correctly', () => {
    // Arrange

    // Act
    const { container } = render(
      <AddBook />,
    );
    // Assert
    expect(container).toMatchSnapshot();
  });

  test('Have the text Add Book ', () => {
    // Arrange

    // Act
    const { getByText } = render(<AddBook />);

    // Assert
    expect(getByText('Add New Book').innerHTML).toBe('Add New Book');
  });

  test('AddBook has two input elements', () => {
    const { getByTestId } = render(<AddBook />);
    const title = getByTestId('book-title-input');
    const author = getByTestId('author-title-input');
    expect(title.placeholder).toBe('Book Title');
    expect(author.placeholder).toBe('Author');
  });

  test('AddBook has one submit button', () => {
    const { getAllByText } = render(<AddBook />);
    const button = getAllByText('Add Book');
    expect(button).toHaveLength(1);
  });
});
