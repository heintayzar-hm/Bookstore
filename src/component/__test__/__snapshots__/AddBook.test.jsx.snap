// Vitest Snapshot v1

exports[`AddBook > renders the initial state correctly 1`] = `
<div>
  <article
    data-testid="form-container"
  >
    <h2>
      Add New Book
    </h2>
    <form>
      <input
        data-testid="book-title-input"
        name="title"
        placeholder="Book Title"
        required=""
        type="text"
        value=""
      />
      <input
        data-testid="author-title-input"
        name="author"
        placeholder="Author"
        required=""
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Book
      </button>
    </form>
  </article>
</div>
`;
